### ISSUE TITLE
The tritonus remaining is not included

### ISSUE THREAD

**Issue by GoogleCodeExporter** [pdudits/soundlibs#1]
```
The pom tritonus share doesn't include the remaining plugin.
```

Original issue reported on code.google.com by `hadrien....@gmail.com` on 28 Nov 2012 at 3:08

**Comment by DV8FromTheWorld** [comment]
It has been a long time since this has been looked at, but it is still a problem. The lack of tritonus_remaining in a proper maven repo is quite annoying.

**Comment by pdudits** [comment]
It's been some time since I needed any of these... can you point me to source repo of tritonus_remaining?

**Comment by pdudits** [comment]
Ah the config is here https://github.com/pdudits/soundlibs/blob/master/tritonus-cvs/build-common.xml#L251

Yes, looks like I could add that. Do you volunteer to try it when I put it into staging? Should happen tomorrow, provided I will find my signing key for the repo ;)

**Comment by pdudits** [comment]
All right, the package dependencies within tritonus are a complete mess, so extracting just exactly what's in remaining would be too much work. So I could do tritonus-all with everything that compiles from the source.

**Comment by DV8FromTheWorld** [comment]
Is signing and uploading the plugin as available on the tritonus site with the proper required dependencies in the pom not an option?

**Comment by DV8FromTheWorld** [comment]
However, if you do a tritonus-all you will be to be sure to merge their service configuration files in the META-INF/services folder, otherwise things will overwrite or you will see duplicates.

This is ofc assuming that the other sound things would be included in the [all] like the MP3 SPI

**Comment by pdudits** [comment]
Concerning just uploading the original, this is not what I've been after. I want to provide proper artifacts with source jars that would not start throwing ClassNotFoundExceptions when you use them, yet without duplicate classes (this is where current artifacts fail ;))

You are right with the service provider entries. If you check out the project now, the generated tritonus-all contains just the entries that previously were in remaining.

**Comment by DV8FromTheWorld** [comment]
It would be awesome if there were a build of like.. SoundLibs-all. Currently, I have a Gradle build script that searches each dependency for the META-INF/services folder and upon finding it deals with the service files in it, combining every dependency's services into a master file and placing that in the final jar with all dependencies. This allows for service configuration files to not override.

Current situation that I can see happening (currently encountering it):
Developer Joe starts developing a project called A.
A uses gradle or maven to deal with dependencies and marks the MP3 SPI, Tritonus_Share and Tritonus_Remaining as dependencies.
Developer Joe notices that either Tritonus_Remaning or MP3-SPI do not properly load when creating a Shaded jar (FatJar) containing A and its dependencies. This is, of course, because MP3-SPI and Tritonus_Remaining both define the same kind of services, so their service configuration files overwrite eachother when being placed into the FatJar.
To fix this, Developer Joe creates a custom build system to deal with this, creating merged service configuration files during the build for any conflicting services. (This is what I've done so far).
However, Joe wants to put project A on mavenCentral for others to use in their projects.

This is where the problem occurs. Developer Bob starts project B, referencing project A as a dependency. Maven/Gradle downloads project A along with it's 3 dependencies. Developer Bob wishes to make a FatJar of project B. However, Bob doesn't know about the internals of project A, nor that it makes use of Services, so he doesn't know that there is a problem with A's dependencies, thus he doesn't create a special build system to deal with the service conflicts.
Developer B creates a FatJar containing project B, project A, and the 3 dependencies. Now the service configuration files have been overwritten instead of being merged like Joe did, thus the project A has issues, thus project B has issues.

This all would be rectified if project A's dependencies had been merged into a single [all] dependency with proper handling for the service configurations and deployed to a maven repo.

Thus, I would like to propose creating a SoundLibs-all instead of a Tritonus-all. SoundLibs-all would contain all of the SoundLibs. Or, if that doesn't suit your fancy, you could just as easily create a SoundLibs-services which would only encompass the dependencies that contained the META-INF/services folder. The SoundLibs-services pom would then define the dependencies required for all of the projects in the SoundLibs-services artifact.

**Comment by pdudits** [comment]
Thanks for detailed input! I'll try to look into the issue as soon as time permits.

The clean solution would be not to merge, but to properly separate the packages into modules, so they have clean dependencies, and we can really have tritonus-core, tritonus-shared, tritonus-remaining, tritonus-mp3 and so on. Service loaders will look into each file on classpath to find all available services.

If you make a fat jar, then you are right, you need to take care of the service entries, but e. g. [maven-shade-plugin](https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer) can take care of that, and it is the price of doing a fat jar. I can tell you that merging OSGi manifest entries in fat jar is even greater fun (if you're into bitter humor).

Generally soundlibs-all would be the next step after I'd do tritonus-all with all of service entries from https://github.com/pdudits/soundlibs/tree/master/tritonus-cvs/src/packaging that do not require native bindings. Then it's just matter of doing the shade plugin (and fighting so that resulting pom would not have the dependencies to downstream).

**Comment by DV8FromTheWorld** [comment]
I haven't looked into the maven-shade-plugin, how does it deal with the overwriting services issue? Does it go through and merge the service configuration files?

**Comment by DV8FromTheWorld** [comment]
So, after pretty extensive searching, I can say that this is the only repository that provides a proper tritonus_share. In my search for a tritonus_remaining I found a few repositories that had it, however it was compiled against a slightly modified tritonus_shared, specifically they had modified the AudioUtils#millis2Frames(long, double) -> (long, float). Big problem.

Anyways, ignore the tritonus-all or soundlibs-all for a bit, I'd like to offer any help you need in getting tritonus_remaining, along with some of the other tritonus libs, on to maven central.

Also a question: Why is tritonus_shared versioned as 0.3.7-2 instead of 0.3.6? Same question towards the MP3SPI of 1.9.5-1 instead of 1.9.5

**Comment by pdudits** [comment]
I would gladly accept any help with that. Like I said original remaining had lots of dependencies to other packages all around the tritonus source, that's why I went for tritonus-all.

Concerning versions, I'm followed the version present it the source code repository, as I couldn't find the revision exactly corresponding to the released version. The -n suffix represents version of packaging, it case it would need improvement (tritonus would definetely need some ;)).

Please consider trying tritonus-all first (already present in the repo), especially whether service provider section is complete enough.

**Comment by pdudits** [comment]
`tritonus-all` has been now released (in version 0.3.7-1). I believe this is the only thing that we can do without  putting massive effort of cleaning up package dependencies within tritonus itself.
